### PR TITLE
Adapt MapMakerObs to also compute an EDispMap and PSFMap

### PR DESCRIPTION
**Pull request by AtreyeeS**
Hello @adonath @registerrier @luca-giunti ! This PR addresses one of the steps towards dataset stacking.

I had to modify `MapMaker` a bit to ensure tests do not fail. This is mostly a patchwork, as `MapMaker` will go away soon. I am not sure how the `sum_over_axes()` should work for the psf, and edisp maps. Since stackings are being implemented separately, I did not do the stacking of the maps in the MapMaker either.

The CTA DC1 psf files are `gammapy.irf.psf_gauss.EnergyDependentMultiGaussPSF` instead of `gammapy.irf.psf_3d.PSF3D`. I do not know how to handle this, so for now, I have just put a check.

### CODE REVIEW

File: gammapy/cube/make.py [hunk 1]
```diff
@@ -217,16 +233,22 @@ def __init__(
         self.exclusion_mask = exclusion_mask
         self.background_oversampling = background_oversampling
         self.maps = {}
+        self.migra_axis = (
+            migra_axis
+            if migra_axis
+            else MapAxis(nodes=np.linspace(0.0, 1.0, 11), unit="", name="migra")
```
Comment (AtreyeeS) [inline gammapy/cube/make.py]: Are these default values for the `migra` and `rad` axis reasonable?

File: gammapy/cube/make.py [hunk 2]
```diff
@@ -303,10 +332,39 @@ def _make_background(self):
 
         self.maps["background"] = background
 
+    def _make_edisp(self):
+
+        energy_axis = self.geom_true.get_axis_by_name("ENERGY")
+        geom_migra = self.geom_true.to_image().to_cube([self.migra_axis, energy_axis])
```
Comment (AtreyeeS) [inline gammapy/cube/make.py]: This is a bit annoying, doing a `to_image()` and then a `to_cube()`. But the `EdispMap` class needs the first axis to be `migra`, and the second, `energy`.

File: gammapy/cube/make.py [hunk 3]
```diff
@@ -209,6 +223,8 @@ def __init__(
         geom_true=None,
```
Comment (AtreyeeS) [inline gammapy/cube/make.py]: There are basically 3 geoms involved here:

1. [e_reco, spatial_fine] - The geom for counts and background - `geom`
2. [e_true, spatial_coarse] - The geom for IRFs - `geom_true`
3. [e_true, spatial_fine] - The final exposure geom. This is created on the fly

File: gammapy/cube/make.py [hunk 4]
```diff
@@ -303,10 +332,39 @@ def _make_background(self):
...
+            geom=geom_migra,
+            max_offset=self.offset_max,
+            exposure_map=self.maps["exposure_irf"],
+        )
+        self.maps["edisp"] = edisp_map
+
+    def _make_psf(self):
+        # CTA 1DC psf is gammapy.irf.psf_gauss.EnergyDependentMultiGaussPSF
+        # TODO: correct handling for 1DC data?
+        if type(self.observation.psf) is not PSF3D:
```
Comment (adonath) [inline gammapy/cube/make.py]: For this kind of checks one typically uses `if isinstance(..., PSF3D)`. Note that there is https://github.com/gammapy/gammapy/blob/master/gammapy/irf/psf_gauss.py#L451, which allows you to transform to a `PSF3D`.

File: gammapy/cube/make.py (hunk 3 above)
Comment (adonath) [reply gammapy/cube/make.py]: This makes sense. I guess alternatively one could change to the parametrisation `geom, energy_axis_true, rad_axis, migra_axis, bins_irf, margin_irf`. Which is a bit simpler, because users don't have to create two geometries. But I'm fine to keep as is...

File: gammapy/cube/make.py (hunk 1 above)
Comment (adonath) [reply gammapy/cube/make.py]: Maybe declare the default `migra_axis` and `rad_axis` at the top of the file as `RAD_AXIS_DEFAULT` and `MIGRA_AXIS_DEFAULT` os they can be imported? I'm pretty sure we'll need these defaults in other places as well (e.g. `MapDataset.create()`)...

I'm not sure what good default are, I think we've never done areal study how good the edisp and psf should be sampled. As far as I know the default values for the FITS data export are:

```
RAD_MAX = 0.6656057834625244
RAD_AXIS_DEFAULT = MapAxis.from_bounds(0, RAD_MAX, nbin=144, node_type="edges", name="theta", unit="deg")
MIGRA_AXIS_DEFAULT = MapAxis.from_bounds(0.2, 5, nbin=160, node_type="edges", name="migra")
``` 
But this seems heavily oversampled...maybe @registerrier has an idea?

File: gammapy/cube/make.py [hunk 5]
```diff
@@ -93,10 +98,11 @@ def run(self, observations, selection=None):
             maps_obs = obs_maker.run(selection)
 
             for name in selection:
-                data = maps_obs[name].quantity.to_value(maps[name].unit)
                 if name == "exposure":
+                    data = maps_obs[name].quantity.to_value(maps[name].unit)
                     maps[name].fill_by_coord(obs_maker.coords_etrue, data)
-                else:
+                elif name == "counts" or name == "background":
+                    data = maps_obs[name].quantity.to_value(maps[name].unit)
```
Comment (adonath) [inline gammapy/cube/make.py]: Why is this needed? I think it's already handled correctly: https://github.com/gammapy/gammapy/blob/master/gammapy/maps/wcsnd.py#L218

File: gammapy/cube/make.py [hunk 6]
```diff
@@ -93,10 +98,11 @@ def run(self, observations, selection=None):
             maps_obs = obs_maker.run(selection)
 
             for name in selection:
-                data = maps_obs[name].quantity.to_value(maps[name].unit)
                 if name == "exposure":
+                    data = maps_obs[name].quantity.to_value(maps[name].unit)
```
Comment (adonath) [inline gammapy/cube/make.py]: Why is this needed? I think it's already handled correctly: https://github.com/gammapy/gammapy/blob/master/gammapy/maps/wcsnd.py#L218

File: gammapy/cube/make.py [hunk 7]
```diff
@@ -93,10 +98,11 @@ def run(self, observations, selection=None):
...
                     maps[name].fill_by_coord(obs_maker.coords_etrue, data)
-                else:
+                elif name == "counts" or name == "background":
```
Comment (adonath) [inline gammapy/cube/make.py]: Ok, to just omit the stacking of the `psf` and `edisp`, but maybe change the default of the `MapMaker.run()` to not compute those maps, otherwise it will slow down our current tutorials and user scripts a lot for no reason.

File: gammapy/cube/make.py [hunk 8]
```diff
@@ -144,6 +150,9 @@ def _maps_sum_over_axes(maps, spectrum, keepdims):
         for name, map in maps.items():
             if name == "exposure":
                 map = _map_spectrum_weight(map, spectrum)
+            if name == "psf" or name == "edisp":
```
Comment (adonath) [inline gammapy/cube/make.py]: OK, to omit this for now, but I guess in general we should reduce the `psf` map as well to only contain one energy bin...the `edisp` can be probably dropped in this case.  Maybe leave a TODO comment here?

File: gammapy/cube/make.py (hunk 4 above)
Comment (AtreyeeS) [reply gammapy/cube/make.py]: Thanks! I add a conversion now.

File: gammapy/cube/make.py (hunk 1 above)
Comment (AtreyeeS) [reply gammapy/cube/make.py]: I discussed a bit with @bkhelifi and have put some new defaults for now. This is relatively coarse, but can be a starting point.

File: gammapy/cube/make.py (hunk 8 above)
Comment (AtreyeeS) [reply gammapy/cube/make.py]: Added a todo

File: gammapy/cube/make.py (hunk 7 above)
Comment (AtreyeeS) [reply gammapy/cube/make.py]: Ok, applied a selection

File: gammapy/cube/make.py (hunk 6 above)
Comment (AtreyeeS) [reply gammapy/cube/make.py]: Changed this by applying selection

File: gammapy/cube/make.py (hunk 3 above)
Comment (AtreyeeS) [reply gammapy/cube/make.py]: We can move to the parametrisation you suggest, specially when we have good defaults for the `e_true` and other axes. I think the user should normally not be interested in IRF binnings. Perhaps we keep this for now and change this once we remove `MapMaker`

**Comment by AtreyeeS** [comment]
Thanks @adonath ! I have addressed the changes you suggested

File: gammapy/cube/make.py (hunk 1 above)
Comment (registerrier) [reply gammapy/cube/make.py]: I would say that the default should the binning used in the IRF. It is safe, it can be too large, but a knowledgeable user could change it.

File: gammapy/cube/make.py [hunk 9]
```diff
@@ -1,17 +1,30 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 import logging
+import numpy as np
```
Comment (registerrier) [inline gammapy/cube/make.py]: remove if unused

File: gammapy/cube/make.py [hunk 10]
```diff
@@ -217,16 +239,14 @@ def __init__(
         self.exclusion_mask = exclusion_mask
         self.background_oversampling = background_oversampling
         self.maps = {}
+        self.migra_axis = migra_axis if migra_axis else MIGRA_AXIS_DEFAULT
```
[the remark this answers is not in the document]
Comment (AtreyeeS) [reply gammapy/cube/make.py]: Ok. Modified to extract the axes from the IRFs.

File: gammapy/cube/make.py [hunk 11]
```diff
@@ -303,10 +325,47 @@ def _make_background(self):
...
+            geom=geom_migra,
+            max_offset=self.offset_max,
+            exposure_map=self.maps["exposure_irf"],
+        )
+        self.maps["edisp"] = edisp_map
+
+    def _make_psf(self):
+        psf = self.observation.psf
+        if isinstance(psf, EnergyDependentMultiGaussPSF):
+            rad = np.linspace(0, 0.66, 67) * u.deg  # Arbitrary binning of 0.01 in rad
```
Comment (AtreyeeS) [inline gammapy/cube/make.py]: @registerrier : Some `rad` values are still needed here. Put in an arbitrary binning here.

Comment (adonath) [reply gammapy/cube/make.py]: One tiny last comment: maybe it's better to introduce the default directly in `EnergyDependentMultiGaussPSF.to_psf3d()`? Then you get rid of the extra code here and it becomes available for other use-cases as well...so you could put `.to_psf3d(rad=None)` and in `to_psf3d()` you put `if rad is None: ...`

Comment (AtreyeeS) [reply gammapy/cube/make.py]: Done!

File: gammapy/cube/psf_map.py [hunk 12]
```diff
@@ -3,7 +3,7 @@
 import astropy.io.fits as fits
 import astropy.units as u
 from astropy.coordinates import Angle
-from gammapy.cube import PSFKernel
+from gammapy.cube.psf_kernel import PSFKernel
```
Comment (registerrier) [inline gammapy/cube/psf_map.py]: Why can't you import from `gammapy.cube` directly?

Comment (AtreyeeS) [reply gammapy/cube/psf_map.py]: I don't understand why, but it was giving an import error.

Comment (registerrier) [reply gammapy/cube/psf_map.py]: Solved in https://github.com/gammapy/gammapy/pull/2358/commits/de85dbd70a9936fa6335e2d3489b283d10027b3f

**Comment by registerrier** [comment]
Issues are solved, Travis passes. Merging.
Thanks @AtreyeeS !
